Tweaks for CWS CUJ "I want to publish my item programmatically" (#1860)
GoogleChrome/developer.chrome.com: site/en/docs/webstore/using_webstore_api/index.md

layout: "layouts/doc-post.njk"
title: "Using the Chrome Web Store Publish API"
#date: TODO
updated: 2021-12-21
description: How to programmatically create, update, and publish items in the Chrome Web Store.
---

## Overview {: #overview }

The Chrome Web Store Publish API provides a set of REST endpoints for programmatically creating,
updating, and publishing items in the Chrome Web Store.

## Before you begin {: #beforeyoubegin }

Take the following steps to get the access keys needed:

**Enable the Chrome Web Store API**.

1. Go to the [Google Cloud Console][google-dev-console].

> 2. Create a new project or select an existing one.

Author: @jpmedley Could I add a screenshot? It might make it easier to follow.

3. In the **search bar** type in “Chrome Web Store API”.
4. Enable the **Chrome Web Store API**.

**Customize a consent screen**.

1. Go to **Credentials**.

> 2. Click on **Configure consent screen**

Reviewer: ```suggestion
2. Click **Configure consent screen**.
```

> 3. Select **External** > **Create**

Reviewer: ```suggestion
3. Select **External** then **Create**.
```

> 4. Fill out the **App information** required fields > **Save and Continue**

Reviewer: ```suggestion
4. Fill out the requred **App information** fields (listed below) then click **Save and Continue**.
```

    - App name.
    - User Support email.
    - Developer contact email.

> 5. Skip Scopes, click **Save** > **Continue**.

Reviewer: ```suggestion
5. Skip Scopes. click **Save** then **Continue**.
```

> 6. Add your email to **Test users**, then **Save** > **Continue**

Reviewer: ```suggestion
6. Add your email to **Test users**, then click **Save** then **Continue**.
```

**Get the access keys**.

1. Go to **Credentials**.

> 2. Click **Create Credentials** > **OAuth client ID**.

Reviewer: ```suggestion
2. Click **Create Credentials** then **OAuth client ID**.
```

4. For **Application type**, choose **Desktop App**.

> 5. Fill out the name, then click **Create**

Reviewer: ```suggestion
5. Fill out the name, then click **Create**.
```

layout: "layouts/doc-post.njk"
title: "Using the Chrome Web Store Publish API"
#date: TODO
updated: 2021-12-21
description: How to programmatically create, update, and publish items in the Chrome Web Store.
---

## Overview {: #overview }

The Chrome Web Store Publish API provides a set of REST endpoints for programmatically creating,
updating, and publishing items in the Chrome Web Store.

## Initial setup {: #setup }

Take the following steps to get the access keys needed:

### Enable the Chrome Web Store API {: #enable-cws-api }

1. Go to the [Google Cloud Console][google-dev-console].
1. Create a new project or select an existing one.
   {% Img src="image/BhuKGJaIeLNPW9ehns59NfwqKxF2/g5o7qpsDv5kKiQmlCwkv.png", alt="Create a new
   project in the Google Console", width="600", height="149" %}
1. In the **search bar** type “Chrome Web Store API”.
1. Enable the **Chrome Web Store API**.

> ### Configure the OAuth consent screen  {: #oauth-setup }

Reviewer 2: @AmySteam, I revised this line to match my suggestion here: https://github.com/GoogleChrome/developer.chrome.com/pull/1860/files/59708cac9a503de00b29ecc9b6ce9d04fa59a550#r774817922 Please resolve this comment if you're comfortable with this change.

layout: "layouts/doc-post.njk"
title: "Using the Chrome Web Store Publish API"
#date: TODO
updated: 2021-12-21
description: How to programmatically create, update, and publish items in the Chrome Web Store.
---

## Overview {: #overview }

The Chrome Web Store Publish API provides a set of REST endpoints for programmatically creating,
updating, and publishing items in the Chrome Web Store.

> ## Before you begin {: #beforeyoubegin }

Reviewer 2: ```suggestion
## Initial setup {: #setup }
```

The title "Before you begin" is a little weird because it's a decent amount of work that you can't skip. I think "Initial setup", "Preparing the way", "API configuration", or something else might be a better framing for what we cover in this section.

> Take the following steps to get the access keys needed:

Reviewer 2: ```suggestion
Before you can begin making REST calls against the Chrome Web Store, you will need to enable the
Chrome Web Store API, configure your OAuth consent screen, and retrieve your API access keys. The
following sections walk through this process.
```

Took a shot at rephrasing the intro to this section to provide a fuller overview of what's to come.

> **Enable the Chrome Web Store API**.

Reviewer 2: ```suggestion
### Enable the Chrome Web Store API {: #enable-cws-api }
```

1. Go to the [Google Cloud Console][google-dev-console].
1. Create a new project or select an existing one.
   {% Img src="image/BhuKGJaIeLNPW9ehns59NfwqKxF2/g5o7qpsDv5kKiQmlCwkv.png", alt="Create a new
   project in the Google Console", width="600", height="149" %}  
1. In the **search bar** type “Chrome Web Store API”.
1. Enable the **Chrome Web Store API**.

> **Customize a consent screen**.

Reviewer 2: ```suggestion
### Configure the OAuth consent screen  {: #oauth-setup }
```

I think this phrasing is a little more accurate. Going through this process isn't just customizing these values, it's setting them up in the first place.

Reviewer 2: We've already merged the other sub-header changes. If I push an update to make this an H3, this line will be lost because the comment will be referencing an out of date line of code. As such, I'm going to push an update to merge this suggestion and tag the current comment so we don't forget to circle back and discuss this text change.

1. Go to **OAuth consent screen**.
1. Select **External** then **Create**.
   {% Img src="image/BhuKGJaIeLNPW9ehns59NfwqKxF2/epcr782dUmXvZwaL5jid.png", alt="Create an Oauth
   consent screen", width="547", height="573" %}
1. Fill out the required **App information** fields (listed below) then click **Save and Continue**.
    - App name.
    - User Support email.
    - Developer contact email.
1. Skip Scopes. click **Save** then **Continue**.

> 1. Add your email to **Test users**, then click **Save** then **Continue**.

Reviewer 2: ```suggestion
1. Add your email address to **Test users**, then click **Save** then **Continue**.
```

> **Get the access keys**.

Reviewer 2: ```suggestion
### Get the access keys {: #get-keys }
```

1. Go to **Credentials**.
1. Click **Create Credentials** then **OAuth client ID**.
  {% Img src="image/BhuKGJaIeLNPW9ehns59NfwqKxF2/vZdPhBUnJytNg5W9TG0I.png", alt="Create credentials", width="657", height="106" %}  
1. For **Application type**, choose **Desktop App**.
1. Fill out the name, then click **Create**.

The console will provide the client ID and client secret. You can retrieve an access token to work
with the API. For example, enter this URL in your browser, replacing the $CLIENT_ID with the one for

> your app:

Reviewer 2: I think it may be best to introduce another section titled "Testing your OAuth application" and to move everything between here and "Using the API" in that area. The last line of the previous section would be "The console will provide the client ID and client secret" and the first sentence of the new section would start with "You can retrieve…"

WDTY, @AmySteam?

Author: I like this suggestion! :)

{% Aside %}

Make sure you are requesting the token using the Google developer Account which owns the Chrome Web
Store items you want to manage. This account can be different from the account you created the

> Google Developers Console project. For example, you can create an application for other developers

Reviewer: ```suggestion
Google Developers Console project with. For example, you can create an application for other developers
```

```json
{
  "access_token" : "ya29...",
  "expires_in" : 3600,
  "refresh_token" : "1/rwn...",
  "scope": "https://www.googleapis.com/auth/chromewebstore",
  "token_type" : "Bearer",
}
```

You can now use the `access_token` to call the API. You can also use the refresh token to get future
access tokens. Note that tokens expire after an hour.

{% Aside %}

For more information about getting OAuth 2.0 access tokens, see [Using OAuth 2.0 to Access Google
APIs][oauth2].

{% endAside %}

## Using the API {: #usingtheapi }

Once you have an access token, your extension can then use the Chrome Web Store Publish API. There
are endpoints for creating a new item, updating an existing item, and publishing an item. 

Below is a list of considerations for using the Publish API:

> - Developers are required to enable Two-Step Verification for their Google Account to publish or

Reviewer: I don't think 'two-step verification' is a proper noun. In any case, please provide a link that describes how to do this.

Reviewer: @dotproto Which of these is the appropriate link?

Reviewer 2: To my knowledge "2-Step Verification" is the prefered term. I think the [How to turn on 2-Step Verification](https://support.google.com/accounts/answer/185839?hl=en&ref_topic=2954345) article is a better reference point than the announcement.

Author: Sounds good, let's go with that.

> - Developers are required to enable Two-Step Verification for their Google Account to publish or

Reviewer: ```suggestion
- Developers are required to enable Two-Step Verification for their Google account to publish or
```

{% Aside %}

For a full description of the insert method, see [Items:Insert][cws-api-insert].

{% endAside %}

### Uploading a package to update an existing store item {: #uploadexisitng }

```text
Endpoint: https://www.googleapis.com/upload/chromewebstore/v1.1/items/$ITEM_ID
Type: PUT
Header Parameters: $TOKEN: the access token
Body content: the package file to upload

> ```

Reviewer 2: I'm not wild about this presentation, but I don't want to bog down this PR with any more stylistic changs than I already have.

Reviewer 2: @jpmedley, do you know if we have an established convention for placeholder values in code snippets? I found the [Placeholders page](https://developers.google.com/style/placeholders#placeholders) in the Google developer documentation style guide, but I don't know if the Markdown guidance applies to dcc or devsite.

Reviewer 2: FYI, I created a feature request for `<var>` support: #/1902

Reviewer: We don't have any good formatting options because we can't do anything inside the backticks. We need to handle this with text. After the example:

Where:
* `ITEM_ID` is ...
* `TOKEN` is ...

If you only have one variable to define, don't use bullets. Make it a single sentence. 

I'm not crazy about the '`$`'. Granting that I don't know anything about the conventions of extensions, my first question is do I have to define or is this a variable that's provided to me? With the dollar sign, it looks like a bash variable.

Reviewer 2: This is a placeholder value for the purposes of documentation. In this case we're not showing JavaScript, so it's not something you'd have to define or something provided to you. Instead, the expectation is that you'd use this information to craft your implementation. For example, I might take this data

> ```text
> Endpoint: https://www.googleapis.com/upload/chromewebstore/v1.1/items/$APP_ID
> Endpoint: https://www.googleapis.com/upload/chromewebstore/v1.1/items/$ITEM_ID
> Type: PUT
> Header Parameters: $TOKEN: the access token
> Body content: the package file to upload
> ```

and use it to write this in my application

```js
class CwsApiManager {
  // OAuth token used to communicate with the CWS API. Set during initial
  // construction.
  token;

  constructor(oauthToken) {/* ... */}

  async uploadItem({id, zipBlob: zip}) {
    let url = `https://www.googleapis.com/upload/chromewebstore/v1.1/items/${id}`;
    let options = {
      method: 'PUT',
      headers: {Authentication: `Bearer ${this.token}`},
      body: zip,
    }
    return fetch(url, options);
  }
}
```

Reviewer 2: I _thought_ our existing convention was to use angle brackets (for example `<PLACEHOLDER>`). I did a quick search and didn't see this pattern in practice.

After a bit more digging I think I found a pervasive bug where we used `<` instead of `&lt;` for placeholder values. For example, the MV2 [Architecture overview page](https://developer.chrome.com/docs/extensions/mv2/architecture-overview/) shows `chrome-extension:///` instead of `chrome-extension://<extensionID>/<pathToFile>`. I'm going to create an issue and start working on a fix for this.

Reviewer: If the Google standard is brackets, we wouldn't follow it because of the danger of being confused for an HTML tag or a web component.

Reviewer 2: [The Google standard](https://developers.google.com/style/placeholders#placeholders) is to decorate the a placeholder name (all caps with underscores separating words) with var tags (for example, `<var>PLACEHOLDER_NAME</var>`). Failing that, the [Placeholders in code blocks](https://developers.google.com/style/placeholders#placeholders-in-code-blocks) section states:

> * In Markdown on a non-DevSite site, wrap the code block in a code fence (```). Inside a code fence, you can't apply formatting like bold or italic.
> 
>     ````md
>     ```
>     PLACEHOLDER_NAME
>     ```
>     ````

I guess for the moment we can go with that guidance and simply avoid decorating placeholder names with angle brackets or other special syntax.

Reviewer: Whatever the Google standard is, you really should be consistent with existing extensions documents whatever that looks like. I should have opened with that. Conforming to the Google standard is something that should be done all at once rather that piecemeal so that a single documentation set can is consistent with itself.

Certain restrictions on language is the only Google writing standard an individual writing team is obligated to. We generally adopt the Google standard because it's more convenient that writing one from scratch.

Reviewer: >Instead, the expectation is that you'd use this information to craft your implementation. For example, I might take this data

Then I think you're way overthinking this. The brackets won't solve the basic issue, which I didn't understand until now. It's because I didn't have both pieces in view. 

1. First name the placeholders in a more meaningful way. For example: YOUR_APP_ID, YOUR_TOKEN, or whatever works for your content.
2. Use those placeholders _identically_ in both locations.

As I said, we're under no obligation to follow the Google style guide to the letter. Frankly, the convention they decided on doesn't get around the obligation to make these things clear in the text, which is always the first rule of writing. If you do that, the brackets don't add any new information.

Reviewer 2: Thanks, @jpmedley. That all makes sense to me.

If you don't mind, at this point I'm inclined to try to leave the placeholders as they are for now and fix them up in a separate PR.

{% Aside %}

For a full description of the insert method, see [Items:Insert][cws-api-insert].

{% endAside %}

### Uploading a package to update an existing store item {: #uploadexisitng }

```text
Endpoint: https://www.googleapis.com/upload/chromewebstore/v1.1/items/$ITEM_ID
Type: PUT

> Header Parameters: $TOKEN: the access token

Reviewer 2: This format is confusing me. For all of the other items the format is basically "key: description", but here it reads like "key: key: description" or "key: header: description". I'm confused, though, because "$TOKEN" is not a header and replacing $TOKEN with the placeholder value also wouldn't be a valid header.

I think it would be best to change this line to be a bit clearer to the reader. I'd be happy to kick around ideas if you like.

Author: @dotproto I'm open to figuring out a better approach. Since you have more experience with API docs, I'm interested to hear your thoughts.
